File: leetcode/q525/hashtable.js
var findMaxLength = function(nums) {
    const indexMap = new Map();
    indexMap.set(0, -1);
    let globalMax = 0;
    let preSum = 0;
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] == 0) {
            preSum -= 1;
        } else {
            preSum += 1;
        }
        if (indexMap.has(preSum)) {
            globalMax = Math.max(globalMax, i - indexMap.get(preSum));
        } else {
            indexMap.set(preSum, i);
        }
    }
    return globalMax;
};

console.log(findMaxLength([0, 1]));

File: leetcode/q525/hashtable.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        indexMap = {0: -1}
        globalMax = 0
        preSum = 0
        for i in range(len(nums)): # you can omit 0 if you start from 0
            if nums[i] == 0:
                preSum -= 1
            else:
                preSum += 1
            if preSum in indexMap:
                globalMax = max(globalMax, i - indexMap[preSum])
            else:
                indexMap[preSum] = i
        return globalMax